keps/77-dynamically-sized-jobs/README.md

```diff
@@ -0,0 +1,286 @@
+
+
+### Creating Workload Slices
+
+The [GenericJob interface](https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/interface.go#L30-L55) will be updated to handle resize operations of jobs.
+
+```golang
+type GenericJob interface {
+  ...
+  ...
+  ResizeJob(wl *kueue.Workload) error
+}
+```
+Jobs implementing the ResizeJob method will create a Workload Slice for every new replica of a job. 
+
+### Pod Scheduling Gates
+
+Inside **raycluster_webhook** implement schedulingGate injection for pods on RayCluster creation time. Which will then be ungated following a similar behavior as to how a job is suspended and then unsuspended in the beginning. When we have a scale up, the new pods will be gated due to the schedulingGates injection in the webhook.
+
+After the creation of each individual Workload Slice and admission of a Workload Slice, the **workload_scheduling_gates_controller** should be in charge of removing the scheduling gates from each pod. We only need to ungate the number of pods to match the number of admitted pods, this should be a counter. We don’t want to accidentally ungate too many pods since race conditions could happen and we also don’t want to double count.
+
+### Garbage Collecting Workload Slices
+
+The logic of folding and deleting would be isolated in this controller that takes a look at the number of pods that were ungated. We don’t necessarily have to say that this Workload Slice belongs to a specific pod. 
+
+1. You increment the ungated counter and pass the UID of the workload you are folding to the parent workload. If we still don’t see the workload being deleted, we at least know it has been counted towards the parent workload and we cannot count it again.  
+2. This UID can be seen in the parent’s workload spec. 
+
+
+## Phases for MVP (alpha)
+
+### Phase 1 - Scale Down
+Scaling down will be the first phase towards MVP because it can be implemented without introducing the Workload Slices. 
+
+Scaling down a RayCluster won’t involve the creation of Workload Slices, instead it’ll involve an update to the current workload, no requeuing.
+
+1. Compare job's PodSet.Count vs Workload.Spec.PodSets[1].Count (worker group) inside the jobframework generic reconciler. 
+2. Call *updateWorkloadToMatchJob()*, this will construct and update the workload and in turn update the PodSet Count field.
+3. Inside the *Update()* method from the *workload_controller* update the workload in cache and in queue. By updating the workload in cache this will update the cluster queue resource usage and by updating the workload in queue this will trigger the scheduler so that it re-assigns the flavors to the already assumed workload and in this way PodSetAssignments will be updated by applying admission based on the new assignments. 
```

Comment:
I'm not understanding the notes about the assumed workloads.

An assumed workload is pretty much the same as an admitted workload (kueue is just waiting for the updated event). There wouldn't be a rescheduling attempt for it.

Reply:
Yes, I guess I was being a bit redundant with the phrasing in this section. This is related to #1852 when an entry for a resize is created, we need to set the `e.status = assumed` and check for this status inside step 5 of the scheduler code (https://github.com/kubernetes-sigs/kueue/blob/536f1ecbbc10d1bd012bd4d8d7c4fb49a7249d28/pkg/scheduler/scheduler.go#L261). Otherwise, it will return an error inside the admit() method, because the workload is already assumed.  We can move the discussion there.

Reply:
ok, that sounds like an implementation detail that shouldn't be part of the KEP.

But also I still don't understand it. I'll review and comment on the implementation PR.

Reply:
Okok will rephrase this section, thanks